feat: Add support for venv management

n-dusan opened the pull request:
Closes #13 
Closes #14 

Added a venv management script built on top of upgrade-python-package script. 
The main goal of the venv management script is to create and/or upgrade a virtual environment.  

- Example of local venv management script usage:
   - `managevenv --envs-home .\Envs\ --requirements oll-partners-us-dc~=2.12.6 --test --cloudsmith-url <CLOUDSMITH_TOKEN> --auto-upgrade`. 

Expected output is a new venv called `oll-partners-us-dc~=2.12.6` created inside of relative `.\Envs` directory.

Added a find-compatible-versions script which for a given venv, determines whether a new compatible update is available. 

- Example usage:

  - `find-compatible-version --requirements-file D:\library\partners\cityofsanmateo\law-xml\requirements.txt --venv-path .\Envs\oll-partners~=2.12.0\ --cloudsmith-url <CLOUDSMITH_URL> --test`
  
Since I'm calling the upgrade-python-package script from the venv management script as a subprocess, that means that I need the latest version of upgrade-python-package. For local testing I've done that by hardcoding my local path to the upgrade-python-package. Need to resolve since the testing framework won't work. 

- [x] Fix CI tests
- [x] Test out blue/green deployment?
- [x] Add filter versions tests?
- [x] Add changelog markdown file

n-dusan committed:
refact: support 3.6 by not relying on pypi dependencies

We used Pillow and other PyPI dependencies in our testing framework
which made it difficult to test against different versions of Python.
We can stop relying on PyPI altogether and only test against
oll-dependency1, oll-dependency2 and oll-top-level-dependency.
Enabled Python 3.6 in GitHub Actions.

n-dusan committed:
refact: use newer versions of `python` and `checkout` in .github

n-dusan committed:
fix(ci): downgrade ubuntu version

Since Python 3.6 isn't supported on ubuntu-latest, downgrade to ubuntu 20
on GitHub Actions.

n-dusan committed:
feat(ci): test against py 3.11

n-dusan committed:
refact: move cloudsmith validation into `validations.py`

n-dusan committed:
feat(venv): initial pass of venv manager arguments

n-dusan committed:
refact(tests): move upgrade tests into `upgrade_package` dir

New venv management tests will go into a separate testing directory

n-dusan committed:
feat(ci): add macos,windows to testing framework

n-dusan committed:
feat(build): add `packaging` dependency

Dependency resolution filters were ported to an open source dependency called `packaging`. Previously, to parse requirements.txt versions pip relied on setuptools. On newer Python resolution logic is now available to use as part of `packaging`

n-dusan committed:
refact: use `head` to check if `index-url` is valid

Expect `requests` module to exist, as it's a prerequisite to running
`upgrade-python-package` in `setup.cfg`

n-dusan committed:
fix: remove unused import

n-dusan committed:
feat: add util functions for working with venv

n-dusan committed:
feat: parse `requirements.txt` in venv manager

n-dusan committed:
feat: initial work on `create_venv`

n-dusan committed:
feat: determine whether an upgrade is necessary

There are two possible ways of determining whether an upgrade is
available. The first is to manually create a backup venv and
install a dependency in venv with `--upgrade` and then `pip check`.
However, running pip install to determine whether an upgrade is
available is expensive in backup venv is expensive.

The second approach is to rely on pip tooling that determines
whether an upgrade is available. However, to get the list of available
upgrades, we query the index.html page to get dependency candidates.
We can parse the html page to get dependency candidates,
as the PEP that proposes HTML standardization for packages
is finished and implemented in PEP 503 - https://peps.python.org/pep-0503/
Rely on lxml to parse the HTML contents into a tree.

After dependency candidates are parsed from anchor tags, we use
`packaging` `SpecifierSet` to filter out only the versions that are
compatible to the specifier. For examples, please see the docstring in
`_filter_versions`.

n-dusan committed:
chore: formatting

n-dusan committed:
refact: rely on vendored pip packages

Instead of adding dependencies that are necessary/required to parse
wheels, parse requirements or make requests, pip vendors the
dependencies that are part of the pip package. Vendored packages MUST
NOT be modified except as required to successfully vendor them - [1].
This means that we delegate all package-handling logic to packages that
are bundled together with pip. The added benefit of using vendored
packages is that we don't have to manage them ourselves, depending on
which pip version gets installed.
Refactor upgrade script to use vendored `requests` and `packaging`.
This should hopefully dramatically simplify the upgrade-python-package
issue with adding more dependencies to the script.

[1] - https://pip.pypa.io/en/stable/development/vendoring-policy/

n-dusan committed:
feat: install `upgrade-python-package` in newly created venv

- the script will now get installed in every created venv. This allows
  us to run upgrade-python-package as a subprocess in the venv.

  The reason we run upgrade-python-package as subprocess is because the
  script expects the `sys.executable` to already point to the venv.
  Alternatively, we could've changed the `sys.executable` path to the new
  venv path. However, that solution gets potentially complicated, since
  the upgrade script also performs module imports to get constraints
  file. That would mean we would also need to update `sys.path` and
  `sys.prefix` to point to the newly created venv `site-packages`. At
  that point, managing the sys.executables and sys.paths for all the
  different states venvs could be in could cause unpredicted issues.

- create venv without pip, then ensurepip. This resolves permission
  errors that we were getting during upgrade-python-package package
  install on Windows.

n-dusan committed:
fix: update docstrings, return absolute path to venv

n-dusan committed:
feat: add initial testing framework for creating venvs

n-dusan committed:
fix: keep venv until next test execution

n-dusan committed:
test: setup `venv` once per test suite

Use parent `conftest.py` to setup a venv from which all tests will be run

n-dusan committed:
test: fix expected files depending on OS

n-dusan committed:
refact: simplify requirement parse

After testing on Python 3.6, using vendored pip
`packaging` dependency works for older versions of Python.

n-dusan committed:
feat: add temp-venv manager

n-dusan committed:
feat: further work on `temp_venv` context manager

The context manager is a backup venv on which we run the upgrade script.
If the upgrade is successful, then the backup venv becomes the primary
venv, otherwise it gets cleaned up.

n-dusan committed:
feat: rotate backup venv to main venv if upgrade is successful

n-dusan committed:
fix: clear `response_err` if upgrade is successful

Response error represents all of the logging that was done during
upgrade.

n-dusan committed:
wip

n-dusan committed:
fix(upgrade): catch `TypeError` on finding constraints

If oll cannot be imported from venv, we still want to find the
constraints. `TypeError` only happens if the wheels are getting
installed locally.

n-dusan committed:
refact(upgrade): determine compatible local wheel version

Move `_filter_versions` function from venv manager to the upgrade
-python-package script. We should now be able to find the most
compatible local wheel given a set of version requirements. An example
would be if the version specifier set is ~=2.0.0, then the most
compatible version of a package should be ==2.0.2 for local wheels. This
was previously not the case, and resolves issues with upgrading local
dependencies.

n-dusan committed:
refact(managevenv): upgrade local wheels in venv management script

The script needs to be able to upgrade from local wheels in case of
local testing. Add support by specifying wheel path and a wheel update
boolean. Since the flags are only currently used for testing framework,
conditionally include them when invoking the upgrade script.

n-dusan committed:
fix(upgrade): fix local upgrade to return upgrade result

Necessary for local testing for venv management.

n-dusan committed:
fix(test): rename `oll-test-top-level` from `2.0.2` to `2.1.0`

Necessary because we consider the previous `2.0.2` version incompatible
during upgrade tests. To make it more clear, rename the version to
`2.1.0` and update existing testing framework.

n-dusan committed:
refact(test): rename `managevenv` new vens, add remaining test cases

n-dusan committed:
chore(venv): formatting

n-dusan committed:
feat(test): add `managevenv` auto upgrade test cases

Create two initial venvs programatically. The first venv should
transition to newer version of `oll-test-top-level`. The second venv
should remain unchanged. Add test cases demonstrating this behavior.

n-dusan committed:
refact(managevenv): b/g switch venvs if upgrade is successful

Instead of using os-rename or generating uuids, move the active venv to
a temporary one, and move backup venv to be the active venv.

n-dusan committed:
refact(upgrade): rename `responseErr` to `responseOutput`

Naming it `responseOutput` makes it clear that we are storing the output
of upgrade, not just the error messages.

n-dusan committed:
chore(managevenv): add type hints

n-dusan committed:
refact(managevenv): move `requirements` parsing into separate file

This way we can re-use the logic in multiple scripts.

n-dusan committed:
refact(managevenv): simplify pip by adding py_executable param

Instead of making a `venv_pip` function, re-use the `pip` function
defined in upgrade-python-package, but with an optional way of
specifying custom executable using `py_executable` **kwargs.

n-dusan committed:
feat(managevenv): add `get_compatible_versions` script

The script is supposed to determine whether a compatible package to
upgrade is available. Moved over the logic from `manage_venv.py` into a
standalone script.

To determine whether a compatible upgrade is available, we:
 1. Get currently installed package version in venv.
 2. Parse the index.html content of the package index to get all
    versions.
 3. Use vendored pip's packaging `SpecifierSet` to filter only
    compatible versions.
If there are multiple compatible upgrades, we retrieve the most recent
one.

n-dusan committed:
fix(test): simplify venv tests

Instead of patching to determine compatible upgrade, call the functions
directly.

n-dusan committed:
refact(test): store test venvs in `Environments` directory

n-dusan committed:
fix(managevenv): add missing parameter, remove unnecessary if statement

n-dusan committed:
chore(managevenv): add docstrings

n-dusan committed:
test(managevenv): add tests for determining compatible upgrade version

n-dusan committed:
chore: add script command to `setup.cfg`

n-dusan committed:
feat(managevenv): add support for specifying requirements directly

Instead of parsing a requirements.txt file, send requirements as a flag.
Optionally, if the dependency is not specified, parse the
`requirements.txt` file.

n-dusan committed:
feat(managevenv): specify additional dependencies to upgrade

Any additional packages that are supposed to be upgraded, such as `cls`,
can be specified using `--additional-dependencies`.

n-dusan committed:
feat(managevenv): add response status enum to compatible versions script

n-dusan committed:
feat(managevenv): add blue/green deployment flag

We do blue/green deployment of virtual environments. If blue/green
deployment is enabled, we expect the following venv directory structure:
- <ENVS_HOME_DIR>
   - dependency~=2.0.0_blue
   - dependency~=2.0.0_green

Management of switching the blue and green venvs is managed outside of
the script scope. If blue/green flag is disabled, we switch the upgraded_green
to venv _blue automatically.

n-dusan committed:
refact(managevenv): add response states to venv management

n-dusan committed:
fix(test): move `sys.executable` default into function

Testing framework patches the executable path to a test venv. Since
defaulting to sys.executable in function signature uses the default
global Python, rely on specifying a custom executable in function body
instead. Fix tests to not rely on success, since success is determined
by something being installed. Should not matter for test cases since we
are asserting the state of the venv.

n-dusan committed:
test(managevenv): specify upgrade-python script version externally

We should now be able to specify an exact version of
upgrade-python-package script to be used in the new venvs. This allows
us to pin venvs to use a specific version of upgrade script, instead of
always using the latest one. Reason for pinning is that we don't want to
immediately transition to new upgrade-python-package script in venvs,
since a new release could potentially break the venvs.

This also enables us to patch the testing framework so venv management
tests should now work.

n-dusan committed:
fix(managevenv): set log location

n-dusan committed:
fix(test): set log location in tests

n-dusan committed:
fix(upgrade): use requests instead of vendored pip requests

n-dusan committed:
fix(test): do not use cloudsmith url in testing framework

n-dusan committed:
fix(ci): do not run ci on outdated python versions

n-dusan committed:
fix(ci): use latest images, add python 3.12

n-dusan committed:
fix(tests): get tests working for linux

n-dusan committed:
fix(test): get sitepackages if only one sitepackage

This fix is a result for failing linux CI builds, where getsitepackages returned a list with only 1 element, thus raising list index out of range for index [1].

n-dusan committed:
fix: default `check` to True, but use kwargs

n-dusan committed:
refact(build): support Python 3.12, drop support for deprecated Python versions

n-dusan committed:
fix(managevenv): set `--pre` on development index correctly

Also fix a comment for local testing

n-dusan committed:
refact(managevenv): simplify blue venv name

Since we know that the blue venv is the original one,
instead of programatically appending `_blue` to the original venv, we
expect the blue venv to be the original one. The upgraded venv remains
to have a `_green` suffix to the venv name.

n-dusan committed:
test(managevenv): deduplicate venv installed version assertions

Move over the code to a separate function in new `test_utils.py`
function.

n-dusan committed:
test(managevenv): test blue/green venvs

n-dusan committed:
test(upgrade): add unit tests for `filter_versions`

n-dusan committed:
feat: add changelog file

Added changelog with all the changes for previous releases.

n-dusan committed:
feat: add changelog with newest unreleased changes

n-dusan committed:
chore(upgrade): remove unused comment

n-dusan committed:
fix: upgrade dependencies even if pip check does not fail

In venv management script the `--update-all` flag was missing,
indicating that the upgrade won't get applied even if pip install
--no-deps fails. To resolve, we want to specify the `--version` flag,
but we also want to run pip with `--upgrade`, so we can get the latest
point release versions.

renatav commented on CHANGELOG.md:
```diff
@@ -0,0 +1,129 @@
+[#6]: https://github.com/openlawlibrary/upgrade-python-package/pull/6
+
+[Unreleased]:  https://github.com/openlawlibrary/upgrade-python-package/compare/0.6.0...HEAD
+[0.6.0]: https://github.com/openlawlibrary/upgrade-python-package/tree/0.6.0
```
I think that all of this links should include `compare/v1...v2`, like https://github.com/openlawlibrary/upgrade-python-package/compare/0.5.0...0.6.0

n-dusan replied:
Yes, thanks!

renatav commented on setup.cfg:
```diff
@@ -14,24 +14,25 @@ classifiers =
     Natural Language :: English
     Programming Language :: Python :: 3
-    Programming Language :: Python :: 3.6
-    Programming Language :: Python :: 3.7
```
Are we also dropping support for Python 3.7? If so, let's mention it in the changelog  "Drop Python 3.6 support, add support for Python 3.11, 3.12"

n-dusan replied:
Noted

renatav commented on upgrade/scripts/find_compatible_versions.py:
```diff
@@ -0,0 +1,191 @@
+
+from upgrade.scripts.requirements import parse_requirements_txt, to_requirements_obj
+from upgrade.scripts.upgrade_python_package import (
+    filter_versions,
```
Since `upgrade_python_package` is the main module, I think that these functions that should be imported elsewhere should be moved to some other modules in order to avoid accidental circular imports in the future.

n-dusan replied:
Makes sense, I've now pushed the changes in [this](https://github.com/openlawlibrary/upgrade-python-package/pull/15/commits/8082136854d023c105d255b33b18b2f542c88a33) and [this](https://github.com/openlawlibrary/upgrade-python-package/pull/15/commits/bba658cfb540b241b6186dff59c5f41a62f9123d) commit. How does it look?

renatav replied:
Looks good

renatav commented on upgrade/scripts/find_compatible_versions.py:
```diff
@@ -0,0 +1,191 @@
+    parsed_packages_versions = [
+        parse_wheel_filename(tag_el.text)[1] for tag_el in anchor_tags_el
+    ]
+    compatible_versions = filter_versions(
```
Since logging is already set up, might be useful to log this too, for debugging purposes

n-dusan replied:
Added

renatav commented on upgrade/scripts/find_compatible_versions.py:
```diff
@@ -0,0 +1,191 @@
+    installed_version = get_installed_version(requirements_obj, venv_executable)
+    if not installed_version:
+        raise Exception(f"Package {requirements_obj.name} is not installed")
+    print(f"installed_version: {installed_version}")
```
Maybe log instead of calling print? Or was this code just used for debugging purposes?

n-dusan replied:
Added logging info instead of print

renatav commented on upgrade/scripts/find_compatible_versions.py:
```diff
@@ -0,0 +1,191 @@
+    "--test",
+    action="store_true",
+    help="Determines whether log messages will be output to stdout "
+    + "or written to a log file",
```
and is used to determine the debug level

renatav commented on upgrade/scripts/find_compatible_versions.py:
```diff
@@ -0,0 +1,191 @@
+    return None
+
+
+def find_compatible_versions(
```
If my understanding is correct, this will raise an error if the package is not already installed. Could you add a docstring which would mention this? And also why we don't want to return something in that case

n-dusan replied:
Added docstring

renatav commented on upgrade/scripts/manage_venv.py:
```diff
@@ -0,0 +1,430 @@
+from typing import Any, List, Optional
+
+from upgrade.scripts.requirements import parse_requirements_txt, to_requirements_obj
+from upgrade.scripts.upgrade_python_package import is_development_cloudsmith, pip, run
```
Same here, I think that these should be moved out of `upgrade_python_package` if they are needed in multiple places

n-dusan replied:
Moved out of `upgrade_python_package`.

renatav commented on upgrade/scripts/manage_venv.py:
```diff
@@ -0,0 +1,430 @@
+    if upgrade_python_package_version:
+        upgrade_python_package += f"=={upgrade_python_package_version}"
+    try:
+        # # # NOTE: for local testing of unreleased upgrade-python-package,
```
Would it possible to add a flag, like `--local` which would automatically install with `-e` and determine the location of the project assuming that this file is inside it? (not critical)

n-dusan replied:
Added, thanks!

renatav commented on upgrade/scripts/manage_venv.py:
```diff
@@ -0,0 +1,430 @@
+    """Build and upgrade a virtualenv."""
+    venv_path = str(_get_venv_path(envs_home, requirements))
+    error_message = None
+    if not Path(venv_path).exists():
```
Do you think that it could be possible that the directory exists, but is empty/not fully initialized?

n-dusan replied:
Definitely, although we handle and cleanup the venv in Codify (so outside of the scope of this script). I'm not quite sure how to handle that specific inconsistent state though.

n-dusan replied:
Refactored to check if venv state is consistent

renatav commented on upgrade/scripts/manage_venv.py:
```diff
@@ -0,0 +1,430 @@
+
+@contextmanager
+def temporary_upgrade_venv(venv_path: str, blue_green_deployment: bool) -> str:
+    """Create a temporary virtualenv and return the path to the python executable."""
```
Could you describe this and blue-green deployment implementation in more detail?

n-dusan replied:
I've expanded on the docstring a bit

renatav commented on upgrade/scripts/upgrade_python_package.py:
```diff
@@ -256,7 +279,7 @@ def install_wheel(
             # revert back to old package version
             if version is not None:
-                package_name = package_name.split("==")[0]
+                package_name = package_name.split("==")[0] # TODO: why ==?
```
TODO left here. I'm not sure why `==`. I believe that we used to only provide input with `==` in the past, we did not need to support anything else

n-dusan replied:
Ah, that TODO is in a previous commit history, I've removed the TODO. I don't think we need to resolve this immediately, so can create an issue

renatav commented on upgrade/scripts/upgrade_python_package.py:
```diff
@@ -359,13 +377,20 @@ def attempt_to_install_version(
+    args = tuple(arg for arg in pip_args)
     try:
         resp = install_wheel(
             package_install_cmd,
```
I'd stick with kwargs here

n-dusan replied:
Right, my line of thinking was to prepare the args for the expression that's used [here](https://github.com/openlawlibrary/upgrade-python-package/blob/2fcc5b885423aaf411fcd19530f7d35a9a63a71f/upgrade/scripts/upgrade_python_package.py#L221-L222) in `install_wheel`.  We already do the same in `attempt_upgrade` [here](https://github.com/openlawlibrary/upgrade-python-package/blob/2fcc5b885423aaf411fcd19530f7d35a9a63a71f/upgrade/scripts/upgrade_python_package.py#L392-L393). Is there a better way to specify these types of args?

renatav commented on upgrade/tests/data/dependency1/setup.cfg:
```diff
@@ -13,6 +13,7 @@ classifiers =
     License :: OSI Approved :: Apache Software License
     Natural Language :: English
     Programming Language :: Python :: 3
+    Programming Language :: Python :: 3.6
```
Since Python 3.6 and Python 3.7 were removed and 3.10, 3.11 and 3.12 were added, should these be updated too?

n-dusan replied:
Yes, thank you

renatav commented on upgrade/tests/manage_venv/conftest.py:
```diff
@@ -0,0 +1,136 @@
+        yield
+
+
+def install_upgrade_python_package(venv_executable, upgrade_python_package_version):
```
`upgrade_python_package_version` is not used here

n-dusan replied:
Resolved in [this commit](https://github.com/openlawlibrary/upgrade-python-package/pull/15/commits/64dbda92ec98e95986bda1e6fe17dc2491e5ec97).

renatav commented on upgrade/tests/upgrade_package/conftest.py:
```diff
@@ -0,0 +1,51 @@
+
+
+@pytest.fixture()
+def mocked_constraints_path():
```
Why is it necessary to mock constraints path and find spec?

n-dusan replied:
This `conftest.py` was copied over from the top level `conftest.py`, so it's the original testing framework that we did a while back ago, and couldn't remember why. I tried removing it but it seems like it was done to get CI (linux) tests working. After trying to get rid of them [in this commit](https://github.com/openlawlibrary/upgrade-python-package/pull/15/commits/db67dcaae49bd9552879d1b772ea37174213cbe1), the linux tests fail. I'm not sure what the difference here is on remote vs on local machine.

renatav replied:
Found the explanation:
* Without `find_spec` mocked, function won't find the installed module
  from the fresh venv, but from existing venv from which pytest was
  invoked. This lead to module not being run, even though if it's valid.
  Mock `find_spec` to result to True, and assert if the module was run
  from unit tests.

No reason to try spend any more time on that then

renatav commented on upgrade/tests/manage_venv/test_get_compatible_upgrade_versions.py:
```diff
@@ -0,0 +1,61 @@
+from upgrade.scripts.requirements import to_requirements_obj
+
+
+def test_get_compatible_version_where_current_installed_version_is_2_0_0_and_compatible_upgrade_version_is_2_0_1_expect_compatible_upgrade_version_returned(
```
Maybe shorten this name by just saying current_is_2_0_0_and_compatible_is_2_0_1 or something like that? So that the name can fit in one line.

renatav commented on upgrade/tests/manage_venv/test_get_compatible_upgrade_versions.py:
```diff
@@ -0,0 +1,61 @@
+    assert actual == expected
+
+
+def test_get_compatible_version_where_current_installed_version_is_2_0_1_and_compatible_upgrade_version_is_2_0_1_expect_no_compatible_upgrade_version(
```
Same

renatav reviewed (commented):
Not tested, but this is looking good. Some minor comments. However, we will need to to test on the dev server or vagrant

n-dusan committed:
fix(docs): fix changelog comparison hyperlinks

n-dusan committed:
fix(docs): include dropped python support in Changelog

n-dusan committed:
refact: move `filter_versions` out of `upgrade_python_package`

It's a function that's used in multiple modules. Move it outside of the
main entrypoint script, so we avoid potential circular imports.

n-dusan committed:
refact: move shared functions into a shared module

Instead of importing from `upgrade_python_package`, move over critical
functions that are used in multiple modules into `utils.py`. Now we
should be independent from importing from `upgrade_python_package`.

n-dusan committed:
refact(managevenv): add logging statements

renatav commented:
Tested on Vagrant. It was hard to get Vagrant up and running, but when I did, it seemed like the script itself still worked

n-dusan committed:
docs(managevenv): add docstring

n-dusan committed:
refact(managevenv): add local upgrade script installation path

Instead of hardcoding local `upgrade-python-package` path in the venv
management script, add an option to specify the local
upgrade-python-package path. This is intended to be used for local
testing purposes

n-dusan committed:
docs: expand on `temporary_upgrade_venv` docstring

n-dusan committed:
fix: update support Python specifiers

n-dusan committed:
test: fix fn signature

This is a function that mocks the original `install_upgrade_python_package` function in managevenv.py. That means that it has to contain the same amount of parameters as the original function. Instead, just use the star '*' expression to resolve.

n-dusan committed:
fix(test): shorten test function names

n-dusan committed:
refact(managevenv): check if both virtual environments consistent

It's possible that a virtual environment for which we want to run the
upgrade on is inconsistent. In that situation, we don't expect the
newly upgraded venv to be consistent either. In both of those cases, we
raise an UpgradeError and let the caller handle the venv states.
